### PR TITLE
HPCC-28944 Construct variable roxie rows directly into output buffer

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Tested via simple queries in Docker Desktop (though code refactored slightly since and I should really test again)

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Something I got distracted by when looking at where and when data was copied/decompressed between being read off the index and being used. To discuss at some point

**Comment by richardkchapman** [comment]
This currently only changes one activity - should change others too

File: roxie/ccd/ccdqueue.ipp [hunk 1]
```diff
@@ -38,12 +38,14 @@ class CDummyMessagePacker : implements IMessagePacker, public CInterface
 {
 protected:
     unsigned lastput;
+    const IOutputMetaData *outputMeta = nullptr;
```
Comment (ghalliday) [inline roxie/ccd/ccdqueue.ipp:41]: Could be a CachedOutputMetaData - to avoid the virtual on isFixedSize()

File: roxie/udplib/udplib.cmake [hunk 2]
```diff
@@ -44,6 +44,7 @@ include_directories (
          ./../../roxie/roxie
          ./../../system/aeron/aeron-client/src/main/cpp 
          ./../../system/aeron/aeron-driver/src/main/c/
+         ./../../rtl/include
```
Comment (ghalliday) [inline roxie/udplib/udplib.cmake:47]: Suggests the structure is wrong...

File: roxie/udplib/udplib.hpp [hunk 3]
```diff
@@ -82,7 +83,7 @@ class UDPLIB_API ServerIdentifier
 
 extern UDPLIB_API ServerIdentifier myNode;
 
-interface IMessagePacker : extends IInterface
+interface IMessagePacker : extends IEngineRowAllocator
```
Comment (ghalliday) [inline roxie/udplib/udplib.hpp:86]: I'm not sure inheritance is the correct structure.  Probably cleaner to have a queryAllocator() virtual method.  May avoid the need for the include and the include path.

File: roxie/udplib/udptrs.cpp [hunk 4]
```diff
@@ -1231,19 +1231,22 @@ class CMessagePacker : implements IMessagePacker, public CInterface
...
     unsigned data_used;
-    void *mem_buffer;
-    unsigned mem_buffer_size;
+    MemoryBuffer mem_buffer;
     unsigned totalSize;
     bool            packed_request;
     MemoryBuffer    metaInfo;
     bool            last_message_done;
     int             queue_number;
+    const IOutputMetaData *outputMeta = nullptr;
```
Comment (ghalliday) [inline roxie/udplib/udptrs.cpp:1240]: CachedOutputMetaData?

File: vcpkg-osx.code-workspace [hunk 5]
```diff
@@ -13,6 +13,8 @@
 			"-DSUPPRESS_REMBED=ON",
 			"-DSUPPRESS_MEMCACHED=ON",
 			"-DSUPPRESS_MONGODBEMBED=ON",
+			"-DUSE_CASSANDRA=OFF",
```
Comment (ghalliday) [inline vcpkg-osx.code-workspace:16]: unrelated, and probably unneeded with recent changes from Gordon.

**Review by ghalliday — COMMENTED** [review]
@richardkchapman finally looked at this.  a few comments.  I'm not sure the structure is quite right.  Also, is it worth the extra complication?

**Comment by richardkchapman** [comment]
Closing as a wontfix as discussed

**Pull request closed without merging**